File: lib/sidereon/gnss/staleness.ex
defmodule Sidereon.GNSS.Staleness do
  @moduledoc """
  Product-staleness graceful degradation for time-varying GNSS products.

  Time-varying products (IONEX vertical-TEC maps, rapid/predicted SP3 orbit and
  clock files) publish with latency and gaps, so the product for the exact
  requested epoch is not always on hand. This is the Elixir surface over the
  `sidereon-core` selection layer: given a SET of already-parsed products and a
  requested epoch (or epoch range), it returns a usable product plus a
  `Sidereon.GNSS.Staleness.StalenessMetadata` describing which source epoch was
  used and how stale it is, falling back to the most-recent product within a
  configurable staleness cap. A request that would rely on a product older than
  the cap fails with a typed error instead of returning data past the cap, so a
  degraded answer is never substituted silently.

  This layer is pure and does no networking: it selects among products the
  caller has already parsed (`Sidereon.GNSS.SP3.load/1`,
  `Sidereon.GNSS.Ionosphere.load_ionex/1`). Fetching the products is a separate,
  per-binding concern.

  ## Degradation paths

    * `:exact` - a product covers the requested epoch; it is returned untouched,
      so the downstream evaluation is bit-for-bit identical to querying it
      directly. Staleness is zero.
    * `:nearest_prior` (SP3) - no product covers the epoch, so the most-recent
      prior product is used as-is, with staleness measured from its last epoch.
    * `:diurnal_shift` (IONEX) - no product covers the requested day, so a prior
      day's grid is advanced by whole days onto the requested epoch (TEC is
      approximately 24-hour periodic). Only the epoch axis moves; grid values are
      unchanged.

  ## Epochs

  Epochs are a `NaiveDateTime` or `{{y, m, d}, {h, min, s}}` tuple, interpreted
  in the product's own time scale (no leap-second shifting), and converted to
  seconds since J2000 via `Sidereon.GNSS.Time`. The IONEX map-epoch axis is
  integer seconds, so an IONEX request must be a whole-second epoch.
  """

  alias Sidereon.GNSS.SP3
  alias Sidereon.GNSS.Time
  alias Sidereon.NIF

  defmodule Policy do
    @moduledoc """
    Staleness cap for product selection.

    A selection that would rely on a product older than `max_staleness_s` fails
    with `{:beyond_staleness_cap, info}` rather than returning data past the cap.
    The default cap is three days, which spans the typical rapid/predicted
    product latency plus a weekend gap.
    """

    @enforce_keys [:max_staleness_s]
    defstruct [:max_staleness_s]

    @type t :: %__MODULE__{max_staleness_s: float()}

    @seconds_per_day 86_400.0
    @default_days 3.0

    @doc "A policy with the cap expressed in days."
    @spec days(number()) :: t()
    def days(days) when is_number(days), do: %__MODULE__{max_staleness_s: days * @seconds_per_day}

    @doc "A policy with the cap expressed in seconds."
    @spec seconds(number()) :: t()
    def seconds(seconds) when is_number(seconds), do: %__MODULE__{max_staleness_s: seconds / 1.0}

    @doc "The default policy: a three-day staleness cap."
    @spec default() :: t()
    def default, do: days(@default_days)
  end

  defmodule StalenessMetadata do
    @moduledoc """
    Structured description of the product staleness behind a selection result.

    Attached to every selection; a degraded result is never produced without it.
    Epoch fields are seconds since the J2000 epoch (2000-01-01 12:00:00 in the
    product's time scale). `staleness_s` is `requested - source` and is never
    negative. `kind` is the degradation path (`:exact`, `:nearest_prior`, or
    `:diurnal_shift`); `staleness_days` is `staleness_s / 86400` (the whole-day
    offset for a diurnal shift).
    """

    @enforce_keys [
      :kind,
      :requested_epoch_j2000_s,
      :source_epoch_j2000_s,
      :staleness_s,
      :staleness_days
    ]
    defstruct [
      :kind,
      :requested_epoch_j2000_s,
      :source_epoch_j2000_s,
      :staleness_s,
      :staleness_days
    ]

    @type kind :: :exact | :nearest_prior | :diurnal_shift

    @type t :: %__MODULE__{
            kind: kind(),
            requested_epoch_j2000_s: float(),
            source_epoch_j2000_s: float(),
            staleness_s: float(),
            staleness_days: float()
          }
  end

  defmodule Sp3Selection do
    @moduledoc """
    A selected SP3 product paired with its staleness metadata.

    `sp3` is a usable `Sidereon.GNSS.SP3` product: for an `:exact` or
    `:nearest_prior` result it is the caller's own input product, so feeding it
    to `Sidereon.GNSS.SP3.position/4` or `Sidereon.GNSS.Positioning.solve/4` is
    bit-for-bit identical to using that product directly. `metadata` reports the
    degradation kind and staleness.
    """

    alias Sidereon.GNSS.Staleness.StalenessMetadata

    @enforce_keys [:sp3, :metadata]
    defstruct [:sp3, :metadata]

    @type t :: %__MODULE__{
            sp3: SP3.t(),
            metadata: StalenessMetadata.t()
          }
  end

  defmodule IonexSelection do
    @moduledoc """
    A selected IONEX product paired with its staleness metadata.

    `handle` is a usable parsed-IONEX reference: for an `:exact` result it is the
    caller's own input handle (so `Sidereon.GNSS.Ionosphere.ionex_slant_delay/7`
    is bit-for-bit identical to using it directly); for a `:diurnal_shift` result
    it is a fresh handle wrapping the whole-day-advanced grid. `metadata` reports
    the degradation kind and staleness.
    """

    alias Sidereon.GNSS.Staleness.StalenessMetadata

    @enforce_keys [:handle, :metadata]
    defstruct [:handle, :metadata]

    @type t :: %__MODULE__{
            handle: reference(),
            metadata: StalenessMetadata.t()
          }
  end

  @typedoc "An epoch as a `NaiveDateTime` or `{{y, m, d}, {h, min, s}}` tuple."
  @type epoch :: NaiveDateTime.t() | tuple()

  @typedoc """
  A typed selection failure. `info` for `:beyond_staleness_cap` is a map with the
  requested/source epochs, the staleness, and the cap, all in J2000 seconds.
  """
  @type selection_error ::
          :empty_product_set
          | {:invalid_range, float(), float()}
          | {:no_prior_product, float()}
          | {:beyond_staleness_cap, map()}
          | {:invalid_product, String.t()}
          | {:invalid_policy, float()}
          | {:overflow, String.t()}

  @doc """
  Select the SP3 product to use for `epoch`, degrading to the most-recent prior
  product within `policy`.

  An `:exact` result covers `epoch`; a `:nearest_prior` result is the best
  in-cap candidate and may end before `epoch` (its coverage gap is the reported
  staleness), so a downstream solve against it can still fail to serve `epoch`.

  `products` is a list of parsed `Sidereon.GNSS.SP3` products. Returns
  `{:ok, %Sidereon.GNSS.Staleness.Sp3Selection{}}` or `{:error, reason}`.
  """
  @spec select_sp3([SP3.t()], epoch(), Policy.t()) ::
          {:ok, Sp3Selection.t()} | {:error, term()}
  def select_sp3(products, epoch, policy \\ Policy.default()) when is_list(products) do
    with {:ok, epoch_s} <- Time.epoch_to_j2000_seconds_fractional(epoch) do
      handles = Enum.map(products, & &1.handle)

      handles
      |> NIF.staleness_select_sp3(epoch_s, policy.max_staleness_s)
      |> decode_sp3_selection(products)
    end
  rescue
    e in ErlangError -> {:error, e.original}
  end

  @doc """
  Select an SP3 product usable across `[start_epoch, end_epoch]`.

  See `select_sp3/3`; this is the range case, where staleness is measured to the
  range end (the most-stale point).
  """
  @spec select_sp3_over_range([SP3.t()], epoch(), epoch(), Policy.t()) ::
          {:ok, Sp3Selection.t()} | {:error, term()}
  def select_sp3_over_range(products, start_epoch, end_epoch, policy \\ Policy.default()) when is_list(products) do
    with {:ok, start_s} <- Time.epoch_to_j2000_seconds_fractional(start_epoch),
         {:ok, end_s} <- Time.epoch_to_j2000_seconds_fractional(end_epoch) do
      handles = Enum.map(products, & &1.handle)

      handles
      |> NIF.staleness_select_sp3_over_range(start_s, end_s, policy.max_staleness_s)
      |> decode_sp3_selection(products)
    end
  rescue
    e in ErlangError -> {:error, e.original}
  end

  @doc """
  Select an IONEX product usable at `epoch`, degrading to a diurnal-shifted prior
  product within `policy`.

  `handles` is a list of parsed-IONEX references from
  `Sidereon.GNSS.Ionosphere.parse_ionex/1` or `load_ionex/1`. The IONEX axis is
  integer seconds, so `epoch` must be a whole-second epoch. Returns
  `{:ok, %Sidereon.GNSS.Staleness.IonexSelection{}}` or `{:error, reason}`.
  """
  @spec select_ionex([reference()], epoch(), Policy.t()) ::
          {:ok, IonexSelection.t()} | {:error, term()}
  def select_ionex(handles, epoch, policy \\ Policy.default()) when is_list(handles) do
    with {:ok, epoch_s} <- Time.epoch_to_j2000_seconds(epoch) do
      handles
      |> NIF.staleness_select_ionex(epoch_s, policy.max_staleness_s)
      |> decode_ionex_selection(handles)
    end
  rescue
    e in ErlangError -> {:error, e.original}
  end

  @doc """
  Select an IONEX product usable across `[start_epoch, end_epoch]`.

  See `select_ionex/3`; this is the range case.
  """
  @spec select_ionex_over_range([reference()], epoch(), epoch(), Policy.t()) ::
          {:ok, IonexSelection.t()} | {:error, term()}
  def select_ionex_over_range(handles, start_epoch, end_epoch, policy \\ Policy.default()) when is_list(handles) do
    with {:ok, start_s} <- Time.epoch_to_j2000_seconds(start_epoch),
         {:ok, end_s} <- Time.epoch_to_j2000_seconds(end_epoch) do
      handles
      |> NIF.staleness_select_ionex_over_range(start_s, end_s, policy.max_staleness_s)
      |> decode_ionex_selection(handles)
    end
  rescue
    e in ErlangError -> {:error, e.original}
  end

  # --- decoding (shared with the precise-to-broadcast fallback source) ------

  @doc false
  @spec decode_metadata(tuple()) :: StalenessMetadata.t()
  def decode_metadata({kind, requested_s, source_s, staleness_s, staleness_days}) do
    %StalenessMetadata{
      kind: kind,
      requested_epoch_j2000_s: requested_s,
      source_epoch_j2000_s: source_s,
      staleness_s: staleness_s,
      staleness_days: staleness_days
    }
  end

  @doc false
  @spec decode_selection_error(term()) :: selection_error()
  def decode_selection_error(:empty_product_set), do: :empty_product_set
  def decode_selection_error({:invalid_range, _start, _end} = error), do: error
  def decode_selection_error({:no_prior_product, _requested} = error), do: error

  def decode_selection_error({:beyond_staleness_cap, requested, source, staleness, max}) do
    {:beyond_staleness_cap,
     %{
       requested_epoch_j2000_s: requested,
       source_epoch_j2000_s: source,
       staleness_s: staleness,
       max_staleness_s: max
     }}
  end

  def decode_selection_error({:invalid_product, _message} = error), do: error
  def decode_selection_error({:invalid_policy, _max} = error), do: error
  def decode_selection_error({:overflow, _context} = error), do: error
  def decode_selection_error(other), do: other

  defp decode_sp3_selection({:ok, {index, metadata}}, products) do
    {:ok, %Sp3Selection{sp3: Enum.at(products, index), metadata: decode_metadata(metadata)}}
  end

  defp decode_sp3_selection({:error, reason}, _products), do: {:error, decode_selection_error(reason)}

  defp decode_ionex_selection({:ok, {selection, metadata}}, handles) do
    handle =
      case selection do
        {:present, index} -> Enum.at(handles, index)
        {:shifted, handle} -> handle
      end

    {:ok, %IonexSelection{handle: handle, metadata: decode_metadata(metadata)}}
  end

  defp decode_ionex_selection({:error, reason}, _handles), do: {:error, decode_selection_error(reason)}
end

File: native/sidereon_nif/src/staleness.rs
//! Rustler boundary for the `sidereon-core` product-staleness selection layer
//! (`sidereon_core::staleness`).
//!
//! This module is **pure glue**: it gathers the caller's already-parsed IONEX /
//! SP3 resource handles, calls the crate's `select_*` functions under a
//! [`StalenessPolicy`], and encodes the [`StalenessMetadata`] plus a usable
//! handle (or an index into the input set) back to Elixir. No selection,
//! degradation, or interpolation math lives here, and there is no networking:
//! fetching the products is a per-binding concern handled elsewhere.
//!
//! ## Result shape
//!
//! - SP3 selection only ever borrows one of the input products (exact or
//!   nearest-prior, never a synthesized product), so the selected product is
//!   returned as its INDEX into the input list. The Elixir wrapper maps that
//!   index back to the original `%Sidereon.GNSS.SP3{}` struct (with its coverage
//!   metadata intact), so the present/exact path is bit-for-bit the caller's own
//!   product.
//! - IONEX selection is either the present product (returned as an index, so the
//!   exact path is the caller's own untouched grid) or a whole-day diurnal-shift
//!   copy (returned as a fresh resource handle). The two cases are tagged
//!   `{:present, index}` / `{:shifted, handle}`.
//!
//! Every success carries the staleness metadata tuple
//! `{kind, requested_epoch_j2000_s, source_epoch_j2000_s, staleness_s,
//! staleness_days}`; a typed `{:error, reason}` is returned otherwise, so a
//! degraded answer is never substituted silently.

use rustler::{Encoder, Env, ResourceArc, Term};

use sidereon_core::atmosphere::ionosphere::Ionex;
use sidereon_core::ephemeris::Sp3;
use sidereon_core::staleness::{
    select_ionex, select_ionex_over_range, select_sp3, select_sp3_over_range, DegradationKind,
    SelectionError, StalenessMetadata, StalenessPolicy,
};

use crate::iono::IonexResource;
use crate::sp3::Sp3Resource;
use crate::spp::atom_from;

/// Encode a [`StalenessMetadata`] as the Elixir tuple
/// `{kind, requested_epoch_j2000_s, source_epoch_j2000_s, staleness_s,
/// staleness_days}`. Shared with the precise-to-broadcast fallback encoder.
pub(crate) fn metadata_term<'a>(env: Env<'a>, m: &StalenessMetadata) -> Term<'a> {
    let kind = match m.kind {
        DegradationKind::Exact => "exact",
        DegradationKind::NearestPrior => "nearest_prior",
        DegradationKind::DiurnalShift => "diurnal_shift",
    };
    (
        atom_from(env, kind),
        m.requested_epoch_j2000_s,
        m.source_epoch_j2000_s,
        m.staleness_s,
        m.staleness_days,
    )
        .encode(env)
}

/// Encode a [`SelectionError`] as a typed Elixir reason term. Shared with the
/// fallback encoder, where a declined precise selection surfaces as
/// `{:precise_unavailable, reason}`.
pub(crate) fn selection_error_term<'a>(env: Env<'a>, error: &SelectionError) -> Term<'a> {
    match error {
        SelectionError::EmptyProductSet => atom_from(env, "empty_product_set"),
        SelectionError::InvalidRange {
            start_epoch_j2000_s,
            end_epoch_j2000_s,
        } => (
            atom_from(env, "invalid_range"),
            *start_epoch_j2000_s,
            *end_epoch_j2000_s,
        )
            .encode(env),
        SelectionError::NoPriorProduct {
            requested_epoch_j2000_s,
        } => (atom_from(env, "no_prior_product"), *requested_epoch_j2000_s).encode(env),
        SelectionError::BeyondStalenessCap {
            requested_epoch_j2000_s,
            source_epoch_j2000_s,
            staleness_s,
            max_staleness_s,
        } => (
            atom_from(env, "beyond_staleness_cap"),
            *requested_epoch_j2000_s,
            *source_epoch_j2000_s,
            *staleness_s,
            *max_staleness_s,
        )
            .encode(env),
        SelectionError::InvalidProduct(message) => {
            (atom_from(env, "invalid_product"), message.clone()).encode(env)
        }
        SelectionError::InvalidPolicy { max_staleness_s } => {
            (atom_from(env, "invalid_policy"), *max_staleness_s).encode(env)
        }
        SelectionError::Overflow { context } => (atom_from(env, "overflow"), *context).encode(env),
    }
}

/// `{:error, reason}` for a failed selection.
fn selection_error_result<'a>(env: Env<'a>, error: &SelectionError) -> Term<'a> {
    (
        rustler::types::atom::error(),
        selection_error_term(env, error),
    )
        .encode(env)
}

/// Locate the borrowed product within the local slice, by identity. The crate's
/// `select_sp3` / `select_ionex` exact and nearest-prior paths return a borrow of
/// one of `products`, so a pointer-identity match recovers its index (the same
/// position as the caller's handle list).
fn index_of_sp3(products: &[Sp3], selected: &Sp3) -> Option<usize> {
    products.iter().position(|p| std::ptr::eq(p, selected))
}

fn index_of_ionex(products: &[Ionex], selected: &Ionex) -> Option<usize> {
    products.iter().position(|p| std::ptr::eq(p, selected))
}

/// Select an SP3 product usable at `requested_epoch_j2000_s`, degrading to the
/// most-recent prior product within the staleness cap.
///
/// Returns `{:ok, {index, metadata}}` where `index` is the position of the
/// selected product in `handles`, or `{:error, reason}`.
#[rustler::nif]
fn staleness_select_sp3<'a>(
    env: Env<'a>,
    handles: Vec<ResourceArc<Sp3Resource>>,
    requested_epoch_j2000_s: f64,
    max_staleness_s: f64,
) -> Term<'a> {
    sp3_selection(
        env,
        handles,
        requested_epoch_j2000_s,
        requested_epoch_j2000_s,
        max_staleness_s,
    )
}

/// Select an SP3 product usable across `[start, end]` (J2000 seconds).
#[rustler::nif]
fn staleness_select_sp3_over_range<'a>(
    env: Env<'a>,
    handles: Vec<ResourceArc<Sp3Resource>>,
    start_epoch_j2000_s: f64,
    end_epoch_j2000_s: f64,
    max_staleness_s: f64,
) -> Term<'a> {
    sp3_selection(
        env,
        handles,
        start_epoch_j2000_s,
        end_epoch_j2000_s,
        max_staleness_s,
    )
}

fn sp3_selection<'a>(
    env: Env<'a>,
    handles: Vec<ResourceArc<Sp3Resource>>,
    start_epoch_j2000_s: f64,
    end_epoch_j2000_s: f64,
    max_staleness_s: f64,
) -> Term<'a> {
    let products: Vec<Sp3> = handles.iter().map(|h| h.sp3.clone()).collect();
    let policy = StalenessPolicy::seconds(max_staleness_s);
    let result = if start_epoch_j2000_s == end_epoch_j2000_s {
        select_sp3(&products, start_epoch_j2000_s, policy)
    } else {
        select_sp3_over_range(&products, start_epoch_j2000_s, end_epoch_j2000_s, policy)
    };
    match result {
        Ok(selection) => {
            let index = index_of_sp3(&products, selection.sp3())
                .expect("selected SP3 product is one of the inputs");
            (
                rustler::types::atom::ok(),
                (index, metadata_term(env, &selection.metadata())),
            )
                .encode(env)
        }
        Err(error) => selection_error_result(env, &error),
    }
}

/// Select an IONEX product usable at `requested_epoch_j2000_s` (J2000 seconds,
/// integer for the IONEX map-epoch axis), degrading to a diurnal-shifted prior
/// product within the staleness cap.
///
/// Returns `{:ok, {selection, metadata}}` where `selection` is
/// `{:present, index}` (the caller's untouched grid) or `{:shifted, handle}`
/// (a fresh whole-day diurnal-shift copy), or `{:error, reason}`.
#[rustler::nif]
fn staleness_select_ionex<'a>(
    env: Env<'a>,
    handles: Vec<ResourceArc<IonexResource>>,
    requested_epoch_j2000_s: i64,
    max_staleness_s: f64,
) -> Term<'a> {
    ionex_selection(
        env,
        handles,
        requested_epoch_j2000_s,
        requested_epoch_j2000_s,
        max_staleness_s,
    )
}

/// Select an IONEX product usable across `[start, end]` (J2000 seconds, integer).
#[rustler::nif]
fn staleness_select_ionex_over_range<'a>(
    env: Env<'a>,
    handles: Vec<ResourceArc<IonexResource>>,
    start_epoch_j2000_s: i64,
    end_epoch_j2000_s: i64,
    max_staleness_s: f64,
) -> Term<'a> {
    ionex_selection(
        env,
        handles,
        start_epoch_j2000_s,
        end_epoch_j2000_s,
        max_staleness_s,
    )
}

fn ionex_selection<'a>(
    env: Env<'a>,
    handles: Vec<ResourceArc<IonexResource>>,
    start_epoch_j2000_s: i64,
    end_epoch_j2000_s: i64,
    max_staleness_s: f64,
) -> Term<'a> {
    let products: Vec<Ionex> = handles.iter().map(|h| h.ionex.clone()).collect();
    let policy = StalenessPolicy::seconds(max_staleness_s);
    let result = if start_epoch_j2000_s == end_epoch_j2000_s {
        select_ionex(&products, start_epoch_j2000_s, policy)
    } else {
        select_ionex_over_range(&products, start_epoch_j2000_s, end_epoch_j2000_s, policy)
    };
    match result {
        Ok(selection) => {
            let metadata = selection.metadata();
            // An exact selection borrows one of the inputs, so it is returned as
            // an index (zero-copy, the caller's own grid). A diurnal shift is a
            // synthesized product, so it is returned as a fresh handle.
            let selection_term = match index_of_ionex(&products, selection.ionex()) {
                Some(index) => (atom_from(env, "present"), index).encode(env),
                None => {
                    let handle = ResourceArc::new(IonexResource {
                        ionex: selection.ionex().clone(),
                    });
                    (atom_from(env, "shifted"), handle).encode(env)
                }
            };
            (
                rustler::types::atom::ok(),
                (selection_term, metadata_term(env, &metadata)),
            )
                .encode(env)
        }
        Err(error) => selection_error_result(env, &error),
    }
}
